godot-core/src/lib.rs

```diff
@@ -148,3 +148,40 @@ pub mod private {
         std::io::stdout().flush().expect("flush stdout");
     }
 }
+
+macro_rules! generate_gdextension_api_version {
+    (
+        $(
+            $name:ident => {
+                $($version:literal),* $(,)?
+            }
+        ),* $(,)?
+    ) => {
+        $(
+            $(
+                #[cfg(gdextension_api_version = $version)]
+                #[allow(dead_code)]
+                const $name: &str = $version;
+            )*
+        )*
+    };
+}
+
+// If multiple gdextension_api_version's are found then this will generate several structs with the same
+// name, causing a compile error.
+generate_gdextension_api_version!(
+    GDEXTENSION_FULL_API_VERSION => {
+        "4.0.0",
+        "4.0.1",
+        "4.0.2",
+        "4.0.3",
+        "4.1.0",
+    },
```

Comment:
but then we have no way to specify `4.0.0` specifically. as opposed to `4.0.1` etc. since `4.0` means every version of `4.0` including `4.0.0`, `4.0.1` ..., we could add in the syntax of something like `4.0.x` for "any version of 4.0" but that is even less standard terminology.

Hopefully we wont need to specify patch versions a lot obviously but there was already a breaking patch version released so i wouldn't want us to accidentally make it impossible to specify it.

Reply:
> but then we have no way to specify 4.0.0 specifically. as opposed to 4.0.1 etc. since 4.0 means every version of 4.0 including 4.0.0, 4.0.1 ..., we could add in the syntax of something like 4.0.x for "any version of 4.0" but that is even less standard terminology.

Yes, we would do this by having separate names. I don't know which names would be good here, but we should definitely not use "version" if we mean "version range".
```
// Expect a precise version
#[cfg(gdextension_precise_version = "4.0")]
#[cfg(gdextension_precise_version = "4.0.2")]

// Expect a version range
#[cfg(gdextension_branch = "4.0")]
```

Reply:
as agreed on [discord](https://discord.com/channels/723850269347283004/1111742061729824899) i'll go with `gdextension_api` and `gdextension_exact_api` for now.